### ISSUE TITLE
About LM fusion in ASR

### ISSUE THREAD

**Issue by iamxiaoyubei** [tensorflow/lingvo#87]
I have seen LM fusion mentioned in several ASR papers you have given. You said that you are referring to papers such as Listen Attend and Spell. 

But I have never found out that you have implemented LM fusion. 

In the decoder, your p.fusion inherits the NullFusion class, and NullFusion does nothing. 

So you are not implementing LM fusion in ASR? If so, could you show me it? 

If not, could you tell me what I can do if I want to add a LM fusion?

Thank you!!

**Comment by rprabhavalkar** [comment]
Yes, you're right. At the moment, the open-source version of the code base only provides the NullFusion class. We do not have immediate plans to open-source other LM fusion techniques such as shallow fusion (though this might happen at some point in the future).

That being said, implementing support for shallow fusion would basically require that you implement the FProp and the ComputeLogits with LM methods in a base class of FusionBase. You should be able to use the Babelfish LM layers under tasks/lm for training the LM.

Hope that helps,
Rohit.

**Comment by iamxiaoyubei** [comment]
Thank you very much for your prompt response. I will try it! 

In addition, I would like to ask if I can add LM to beam search process like LAS(listen, attend and spell):

![image](https://user-images.githubusercontent.com/21981293/58534368-b7279d80-821d-11e9-9fff-8cf0c0dd784d.png)


This seems to be more convenient to implement.

**Comment by by2101** [comment]
I am implementing fusion for lingvo, too. However, I am confused with some API.
I see that the searching scores are computed in three steps:

`    softmax_input, fusion_states = self.fusion.FProp(
        theta.fusion, prev_fusion_states, softmax_input, cur_target_info.id,
        cur_target_info.padding)

    logits = self._ComputeLogits(theta, softmax_input)

    logits = self.fusion.ComputeLogitsWithLM(
        fusion_states, logits, is_eval=True)`
For shallow fusion, the first step updates the state of the LM.
Then, the logits of Seq2Seq model are computed.
At last add LM outputs and Seq2Seq model outputs together.

However, I don't know why the input of ComputeLogitsWithLM are fusion state and logits. For shallow fusion, it just sums LM outputs and logits, but not LM states. I don't know how to pass the LM outputs in that function. Could you please help me?
 @rprabhavalkar @ewilderj @iamxiaoyubei 
Thank you!

**Comment by iamxiaoyubei** [comment]
@by2101 
I don't know what is your LM outpus means. But if you need, you can pass the LM outputs to LM states. Then it can be passed to this function.
